Assignment3/Assignment3.py

```diff
@@ -0,0 +1,185 @@
+            return
+
+        @keyword_only
+        def setParams(self, inputCols=None, outputCol=None):
+            kwargs = self._input_kwargs
+            return self._set(**kwargs)
+
+        def _transform(self, dataset):
+            input_cols = self.getInputCols()
+            output_col = self.getOutputCol()
+
+            def days(i):
+                return i * 86400
+
+            def hours(i):
+                return i * 3600
+
+            partition = (
+                Window.partitionBy("batter")
+                .orderBy(col("gamedate").cast("timestamp").cast("long"))
+                .rangeBetween(days(-100), hours(-6))
+            )
+
+            dataset = dataset.withColumn(output_col, sum(input_cols[1]).over(partition))
+
+            return dataset
+
+    # create a transformer that creates a rolling at bats variable
+    class getrollingatbatsTransform(
+        Transformer,
+        HasInputCols,
+        HasOutputCol,
+        DefaultParamsReadable,
+        DefaultParamsWritable,
+    ):
+        @keyword_only
+        def __init__(self, inputCols=None, outputCol=None):
+            super(getrollingatbatsTransform, self).__init__()
+            kwargs = self._input_kwargs
+            self.setParams(**kwargs)
+            return
+
+        @keyword_only
+        def setParams(self, inputCols=None, outputCol=None):
+            kwargs = self._input_kwargs
+            return self._set(**kwargs)
+
+        def _transform(self, dataset):
+            input_cols = self.getInputCols()
+            output_col = self.getOutputCol()
+
+            def days(i):
+                return i * 86400
+
+            def hours(i):
+                return i * 3600
+
+            partition = (
+                Window.partitionBy("batter")
+                .orderBy(col("gamedate").cast("timestamp").cast("long"))
+                .rangeBetween(days(-100), hours(-6))
+            )
+
+            dataset = dataset.withColumn(output_col, sum(input_cols[1]).over(partition))
+
+            return dataset
+
+    # create a transformer that creates a rolling battingaverage variable
+    class getrollingbattingaverageTransform(
+        Transformer,
+        HasInputCols,
+        HasOutputCol,
+        DefaultParamsReadable,
+        DefaultParamsWritable,
+    ):
+        @keyword_only
+        def __init__(self, inputCols=None, outputCol=None):
+            super(getrollingbattingaverageTransform, self).__init__()
+            kwargs = self._input_kwargs
+            self.setParams(**kwargs)
+            return
+
+        @keyword_only
+        def setParams(self, inputCols=None, outputCol=None):
+            kwargs = self._input_kwargs
+            return self._set(**kwargs)
+
+        def _transform(self, dataset):
+            output_col = self.getOutputCol()
+
+            dataset = dataset.withColumn(
+                output_col, col("rollinghits") / col("rollingatbats")
+            )
+
+            return dataset
```

Comment:
I like how you've got this transformer separate. Like we talked about, I wish there was a way to do this without keeping the intermediate rolling hits & atbats vars (maybe there is?)